doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,153 @@
+connection are being read from/written to in different threads
+* SSL objects have lots of APIs that are very specific to TLS and would not fit
+well with QUIC
+* We would have to “pollute” the SSL API with lots of QUIC specific
+functions
+* The “SSL” type name doesn’t fit well with QUIC
+
+
+## Candidate Solution: A generic comms API with SSL compat layer
+
+In this approach we would develop a completely new generic secure
+communications API that would support multiple protocol types, e.g. TLS, DLTS,
+QUIC, etc. We might expect to see separate types for generic “connections”
+and “streams”. Internally to a generic connection we would see protocol
+specific objects. We could choose to expose these internal protocol specific
+objects as part of the API, or we could expose them through generic mechanisms
+(e.g. ctrls, OSSL_PARAM etc).
+
+Some features can be simulated in protocols that don’t have them. For example
+TLS doesn’t have a multi-stream capability - but it could be simulated by
+running resumption handshakes to create a new “stream”.
+
+Additionally an “SSL” compatibility layer would be developed. There could
+be QUIC method objects (e.g. “QUIC_client_method'' or
+“QUIC_server_method”). The compatibility layer:
+
+1. Will call the generic API underneath, e.g. Common functions such as
+“SSL_read” and “SSL_write” would be redirected internally  to the
+equivalent generic connection/stream function as appropriate.
+2. For MVP it will only support 1 stream
+3. Should be possible (eventually) to move from the SSL API to the generic API
+e.g.
+
+    ```
+    OSSL_CONN *SSL_get_connection(SSL *ssl);
+
+    ```
+
+
+The new generic API could be internal only for MVP. We might expect it to have
+a context object type, a connection object type, a stream object type, etc.
+Since it is not TLS, DTLS or QUIC specific, we would expect a different prefix.
+
+Inevitably there will have to be mechanisms for making available protocol
+specific functionality as part of the generic API. An open question is how this
+would be done, e.g.
+
+* Exposing a protocol specific object
+* OSSL_PARAMs
+* Some other params style approach but with ability for complex objects (e.g.
+X509)
+
+A possible data model for this approach can be seen here ([whiteboard
+link](https://jamboard.google.com/d/1AM1g-9eUbGAJ3YR9cEVpH6mrpPHCxd-9DtJMG1hNPlk/edit?usp=meet_whiteboard)):
+
+
+![alt_text](images/image1.png "image_tooltip")
+
+
+Pros
+
+* All protocols look the same to an application. Straight forward to switch
+from one protocol to another
+
+Cons
+
+* Protocol differences will pollute the API
+* Inevitably there will be some features that will work differently (or won’t
+be present at all) between different protocols. Therefore applications swapping
+from one protocol to another may still have to have some protocol specific code
+
+
+## Candidate Solution: A completely new QUIC API with SSL compat layer
```

Comment:
This option seems "least bad" to me.
I remain unconvinced that there's much demand for using QUIC as a protocol while retainining the existing `SSL_*`-based API usage of an application, which seems to be what the "SSL compat layer" would be achieving.  My preference would be to make a QUIC API that is independent of the SSL layer.  As davidben commented at https://github.com/openssl/openssl/pull/17172#issuecomment-990125105 (on HPKE, not QUIC or transport APIs, but still relevant), there are a lot of risks in trying to define an abstraction before you know what the needs are, the design is unlikely to be successful.